src/runmode-dpdk.c

```diff
@@ -0,0 +1,1295 @@
+            .rxmode = {
+                    .mq_mode = ETH_MQ_RX_NONE,
+                    .max_rx_pkt_len = RTE_ETHER_MAX_LEN,
+                    .offloads = 0, // turn every offload off to prevent any packet modification
+            },
+            .txmode = {
+                    .mq_mode = ETH_MQ_TX_NONE,
+                    .offloads = 0,
+            },
+    };
+
+    // configure RX offloads
+    if (dev_info->rx_offload_capa & DEV_RX_OFFLOAD_RSS_HASH) {
+        if (iconf->nb_rx_queues > 1) {
+            SCLogConfig("RSS enabled on %s for %d queues", iconf->iface, iconf->nb_rx_queues);
+            port_conf->rx_adv_conf.rss_conf.rss_key = rss_hkey;
+            port_conf->rx_adv_conf.rss_conf.rss_key_len = RSS_HKEY_LEN;
+            port_conf->rx_adv_conf.rss_conf.rss_hf = ETH_RSS_TCP | ETH_RSS_UDP;
+            port_conf->rxmode.mq_mode = ETH_MQ_RX_RSS;
+        } else {
+            SCLogConfig("RSS not enabled on %s", iconf->iface);
+            port_conf->rx_adv_conf.rss_conf.rss_key = NULL;
+            port_conf->rx_adv_conf.rss_conf.rss_hf = 0;
+        }
+    } else {
+        SCLogConfig("RSS not supported on %s", iconf->iface);
+    }
+
+    if (iconf->checksum_mode == CHECKSUM_VALIDATION_DISABLE) {
+        SCLogConfig("Checksum validation disabled on %s", iconf->iface);
+    } else if (dev_info->rx_offload_capa & DEV_RX_OFFLOAD_CHECKSUM) {
+        if (iconf->checksum_mode == CHECKSUM_VALIDATION_ENABLE &&
+                iconf->flags & DPDK_RX_CHECKSUM_OFFLOAD) {
+            SCLogConfig("IP, TCP and UDP checksum validation enabled and offloaded "
+                        "on %s",
+                    iconf->iface);
+            port_conf->rxmode.offloads |= DEV_RX_OFFLOAD_CHECKSUM;
+        } else if (iconf->checksum_mode == CHECKSUM_VALIDATION_ENABLE &&
+                   !(iconf->flags & DPDK_RX_CHECKSUM_OFFLOAD)) {
+            SCLogConfig("Suricata checksum validation enabled "
+                        "but not offloaded on %s",
+                    iconf->iface);
+        }
+    }
+
+    if (dev_info->tx_offload_capa & DEV_TX_OFFLOAD_MBUF_FAST_FREE) {
+        port_conf->txmode.offloads |= DEV_TX_OFFLOAD_MBUF_FAST_FREE;
+    }
+}
+
+int DeviceConfigureQueues(DPDKIfaceConfig *iconf, const struct rte_eth_dev_info *dev_info,
+        const struct rte_eth_conf *port_conf)
+{
+    int retval;
+    uint16_t mtu_size;
+    uint16_t mbuf_size;
+    struct rte_eth_rxconf rxq_conf;
+    struct rte_eth_txconf txq_conf;
+
+    iconf->pkt_mempool_array = SCCalloc(iconf->nb_rx_queues, sizeof(struct rte_mempool *));
+    if (unlikely(iconf->pkt_mempool_array == NULL)) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Could not allocate memory for packet mempool pointers");
+        SCReturnInt(-ENOMEM);
+    }
+
+    for (uint16_t queue_id = 0; queue_id < iconf->nb_rx_queues; queue_id++) {
+        char mempool_name[64];
+        snprintf(mempool_name, 64, "pktmbuf_pool_p%d_q%d", iconf->port_id, queue_id);
```

Comment:
1. Should this mempool name be a configurable in `suricata.yaml.in` for every interface (like `mempool-size` is)?
2. Do we need mempool's for each port's queue? `port_id` is dynamic with DPDK
too (i.e. it depends on how many physical/vdev ports are there, instead of just config).

For context, I tested this PR with Suricata as `--proc-type=primary`, with
`--vdev=net_ring0` (i.e. DPDK's Ring PMD for a virtual device), and `net_ring0`
as the interface setup in the YAML.

It worked! My app runs as `--proc-type=secondary`  and can
`rte_mempool_lookup` and successfully `rte_eth_tx_burst`
to this vdev.

The only minor issue was that I needed to hardcode support for connecting to
Suricata as a primary using the mempool name `pktmbuf_pool_p%d_q%d`.
I'm just wondering if that can be made into a consistent configurable (such as
`pktmbuf_pool_p%d` perhaps using the order of interfaces in the YAML instead of
DPDK's assigned port_id).
That way, both primary/secondary apps have the mempool-name as a configurable
that is deterministic, allowing for more compatibility overall.
As-is, a secondary DPDK app trying to connect to Suricata as primary just _has
to_ have some hardcoded attempts (since port_id changes).

Reply:
Thank you for the testing. 

Regarding the mempool name - yes, I can make it configurable - I would suggest setting a static prefix followed by the name of the NIC and the queue number. It is true that port_id is dynamic and can be hard to attach to from a secondary app. So in the end it could possibly look something like <configurable_prefix_><NIC PCIe address>_<queue id>. 

>2. Do we need mempool's for each port's queue? port_id is dynamic with DPDK too (i.e. it depends on how many physical/vdev ports are there, instead of just config).

Regarding your second point, I have discussed this with my colleagues, and based on their measurements mempool per queue performs better than a shared mempool (per app/per interface). That is the reason why I opted for this variant.

Reply:
Sure, thank you for your work on this PR!

If performance is better that way, I suppose a `configurable prefix + NIC name / port_id + queue_id` will do. Thanks!